### PR TITLE
[GTK][WPE] Deprecate WebKitConsoleMessage API

### PR DESCRIPTION
**Pull request by carlosgcampos**
#### d671caa0409902d3ac1cc879073c83c5131d41dc
<pre>
[GTK][WPE] Deprecate WebKitConsoleMessage API
<a href="https://bugs.webkit.org/show_bug.cgi?id=247791">https://bugs.webkit.org/show_bug.cgi?id=247791</a>

Reviewed by Michael Catanzaro.

The API seems to be unused so we can also remove the implementation.

* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitConsoleMessage.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitConsoleMessage.h.in:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
(webkit_web_page_class_init):
(webkitWebPageCreate):
(webkitWebPageDidSendConsoleMessage): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestConsoleMessage.cpp: Removed.
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(xhrMessageReceivedCallback):
(testWebContextSecurityFileXHR):
(consoleMessageReceivedCallback): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(pageCreatedCallback):
(consoleMessageSentCallback): Deleted.
* Tools/TestWebKitAPI/glib/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/257010@main">https://commits.webkit.org/257010@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/43fc0353a0cc5dc84be4965b31abc5061281ffea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97521 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107025 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167289 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101478 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7114 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35532 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89919 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103694 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103176 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5328 "Found 1 new test failure: fast/forms/ios/focus-button.html (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84153 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32334 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88993 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75259 "Found 4 new API test failures: /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/console-api, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/js-exception, /WebKitGTK/TestConsoleMessage:/webkit/WebKitConsoleMessage/network-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/778 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/766 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21933 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5579 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44404 "Found 30 new test failures: animations/stop-animation-on-suspend.html, fast/images/animated-heics-draw.html, http/tests/navigation/parsed-iframe-dynamic-form-back-entry.html, http/tests/privateClickMeasurement/send-attribution-conversion-request.html, http/wpt/service-workers/cross-site-navigation-same-cookie-lax.https.html, http/wpt/service-workers/fetch-service-worker-preload-download.https.html, imported/w3c/web-platform-tests/css/css-flexbox/flexbox_flex-1-0-Npercent.html, imported/w3c/web-platform-tests/css/css-masking/clip-path-svg-content/clip-path-no-content-003.svg, imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-rotated-will-change-transform.html, imported/w3c/web-platform-tests/css/css-masking/mask-image/mask-image-3c.html ... (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2384 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2016 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41308 "Passed tests") | |

### CODE REVIEW

**Comment by carlosgcampos** [comment]
It seems the EWS bot doesn't know a test has been removed and still tries to run it. We probably need a clean build, or remove the stale binary somehow.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp [hunk 1]
```diff
@@ -638,35 +639,36 @@ static void testWebContextSecurityFileXHR(WebViewTest* test, gconstpointer)
     test->waitUntilLoadFinished();
 
     GUniquePtr<char> jsonURL(g_strdup_printf("file://%s/simple.json", Test::getResourcesDir().data()));
-    GUniquePtr<char> xhr(g_strdup_printf("var xhr = new XMLHttpRequest; xhr.open(\"GET\", \"%s\"); xhr.send();", jsonURL.get()));
+    GUniquePtr<char> xhr(g_strdup_printf("var xhr = new XMLHttpRequest; xhr.open(\"GET\", \"%s\"); xhr.onreadystatechange = ()=> { if (xhr.readyState == 4) { setTimeout(() => { window.webkit.messageHandlers.xhr.postMessage('DONE'); }, 0)} }; xhr.onerror = () => { window.webkit.messageHandlers.xhr.postMessage('ERROR'); }; xhr.send();", jsonURL.get()));
```
Comment (mcatanzaro) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:642]: Converting this to a multi-line string would aid readability.

**Review by mcatanzaro — APPROVED** [review]
Normally I would ask why are you removing the implementation, in case somebody is using this that we just don't know about, but I see it's because it's using InjectedBundle. OK then: risk vs. reward seems worth it.

**Comment by mcatanzaro** [comment]
> It seems the EWS bot doesn't know a test has been removed and still tries to run it. We probably need a clean build, or remove the stale binary somehow.

So there's no mechanism for removing the old test binary from the build directory. That seems normal: I wouldn't expect CMake to clean up stuff that's no longer built from the build directory.

Tools/glib/api_test_runner.py computes the tests to run by enumerating binaries in the build directory.  Ideally CMake would expect CMake to keep track of which tests it built and run only those, but that would probably require switching to CTest instead of hand-rolling our test runner in python. I don't know anything about CTest.

We remove tests so rarely that it might not be worth spending time to investigate this further.

**Comment by ThomasDevoogdt** [comment]
Why? We did use this API, but we should obviously adapt our code now. No problem with that, but it's unclear to me how to get logging from the console. We need a way to get logging in syslog (not systemd). Do you have any suggestions or threads to read?

**Comment by mcatanzaro** [comment]
Drat. I can't think of any good way to redirect the console messages without this API.

> Why?

See https://lists.webkit.org/pipermail/webkit-dev/2022-August/032329.html

**Comment by mcatanzaro** [comment]
I don't know what to do about this.

**Comment by philn** [comment]
Another sad user over there, https://mastodon.nu/@jonasdn/109732221416203674

**Comment by carlosgcampos** [comment]
We can bring it back if really needed, isn't it possible to get the console messages from javascript? The setting to send messages to stdout should still work too.

**Comment by philn** [comment]
> We can bring it back if really needed, isn't it possible to get the console messages from javascript? 

Seems like monkey-patching is doable: https://stackoverflow.com/questions/19846078/how-to-read-from-chromes-console-in-javascript
Looks awful :)

> The setting to send messages to stdout should still work too.

That setting is useful for debugging yes. But in a real app enabling that and parsing stderr sounds like another footgun.

**Comment by ThomasDevoogdt** [comment]
> The setting to send messages to stdout should still work too.

It is also what we are doing now. But it's not particularly handy to do. The code looks like this:
```c++
        if (contains(line, "CONSOLE")) {
            line = trim(replace(line, "CONSOLE", ""));

            if (contains(line, "ERROR"))
                log_error(replace(line, "ERROR", ""));
            else if (contains(line, "LOG"))
                log_info(replace(line, "LOG", ""));
            else if (contains(line, "INFO"))
                log_info(replace(line, "INFO", ""));
            else
                log_debug(replace(line, "DEBUG", ""));
        }
```
But this does not guarantee that the parsing will do its job with any next release.

**Comment by carlosgcampos** [comment]
In case of bringing it back, I think we should add it as UI process API instead of web extension. Is that ok for everybody using this API?

**Comment by jonasdn-spiideo** [comment]
> In case of bringing it back, I think we should add it as UI process API instead of web extension. Is that ok for everybody using this API?

As long as I can use it from wpe (`wpesrc` in gstreamer) I am happy! Thanks!

**Comment by ThomasDevoogdt** [comment]
> In case of bringing it back, I think we should add it as UI process API instead of web extension. Is that ok for everybody using this API?

Yes, that is ok for me! Thx already!

**Comment by carlosgcampos** [comment]
For now I think I'll bring back the functionality to current API since I deprecated and removed the ipementation assuming nobody was actually using it. Then we can decide what to do for the new API, but I would like to understand the use cases of this API.

**Comment by philn** [comment]
One use-case in wpesrc is to forward these messages to the app's pipeline, using element messages. a UIProcess API for this would fit.

**Comment by JezerM** [comment]
I also do use this API in my own project. If it's being deleted, it would be great if there is at least an alternative way to get the console messages.

> In case of bringing it back, I think we should add it as UI process API instead of web extension. Is that ok for everybody using this API?

Yeah, it would be good. However, I think the web-extension API was already fine, specially in those cases where you may have lots of WebKitWebViews and you would like to differentiate each log for each web_view.

**Comment by jonasdn-spiideo** [comment]
Did we get a resolution to this? Where can I track it, if not here?

**Comment by carlosgcampos** [comment]
I brought it back in https://commits.webkit.org/259535@main

**Comment by jonasdn-spiideo** [comment]
thanks!

**Pull request merged**